File: backend/app/schemas/transporter.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class TransporterCreate(BaseModel):
    name: str
    phone: str | None = None
    vehicle_type: str | None = None
    vehicle_number: str | None = None
    base_city: str | None = None
    notes: str | None = None


class TransporterUpdate(BaseModel):
    name: str | None = None
    phone: str | None = None
    vehicle_type: str | None = None
    vehicle_number: str | None = None
    base_city: str | None = None
    notes: str | None = None
    is_active: bool | None = None


class TransporterResponse(BaseModel):
    id: uuid.UUID
    name: str
    phone: str | None
    vehicle_type: str | None
    vehicle_number: str | None
    base_city: str | None
    avg_cost_per_km: float | None
    avg_spoilage_pct: float | None
    on_time_pct: float | None
    total_trips: int
    rating: float | None
    notes: str | None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}

File: backend-spring/src/main/java/com/nearkart/entity/Transporter.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "transporters")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class Transporter {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(nullable = false)
    private String name;

    private String phone;

    @Column(name = "vehicle_type")
    private String vehicleType;

    @Column(name = "vehicle_number")
    private String vehicleNumber;

    @Column(name = "base_city")
    private String baseCity;

    @Column(name = "avg_cost_per_km", precision = 8, scale = 2)
    private BigDecimal avgCostPerKm;

    @Column(name = "avg_spoilage_pct", precision = 5, scale = 2)
    private BigDecimal avgSpoilagePct;

    @Column(name = "on_time_pct", precision = 5, scale = 2)
    private BigDecimal onTimePct;

    @Column(name = "total_trips")
    @Builder.Default
    private Integer totalTrips = 0;

    @Column(precision = 2, scale = 1)
    private BigDecimal rating;

    @Column(name = "is_active", nullable = false)
    @Builder.Default
    private Boolean isActive = true;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", insertable = false, updatable = false)
    private User user;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}
